src/library_fs.js

```diff
@@ -1017,11 +1047,36 @@ FS.staticInit();
       });
     },
     ftruncate(fd, len) {
+      if (len < 0) {
+        throw new FS.ErrnoError({{{ cDefs.EINVAL }}});
+      }
       var stream = FS.getStreamChecked(fd);
+      var node = stream.node;
       if ((stream.flags & {{{ cDefs.O_ACCMODE }}}) === {{{ cDefs.O_RDONLY}}}) {
         throw new FS.ErrnoError({{{ cDefs.EINVAL }}});
       }
-      FS.truncate(stream.node, len);
+      if (!node.node_ops.setattr && !stream.stream_ops.setattr) {
+        throw new FS.ErrnoError({{{ cDefs.EPERM }}});
+      }
+      if (FS.isDir(node.mode)) {
+        throw new FS.ErrnoError({{{ cDefs.EISDIR }}});
+      }
+      if (!FS.isFile(node.mode)) {
+        throw new FS.ErrnoError({{{ cDefs.EINVAL }}});
+      }
+      var errCode = FS.nodePermissions(node, 'w');
+      if (errCode) {
+        throw new FS.ErrnoError(errCode);
+      }
+      var attrs = {
+        size: len,
+        timestamp: Date.now()
+      };
+      if (stream.stream_ops.setattr) {
+        stream.stream_ops.setattr(stream, attrs);
+      } else {
+        node.node_ops.setattr(node, attrs);
+      }
```

Comment:
Seems like a fair amount of duplication here between `truncate` and `ftruncate`.. is there some way to extract the common stuff?

Reply:
Yeah can do.